soroban-env-host/src/host.rs

```diff
@@ -2697,9 +2705,11 @@ impl VmCallerEnv for Host {
     ) -> Result<BytesObject, HostError> {
         let u = self.u8_from_u32val_input("u", u)?;
         let vnew = self.visit_obj(b, move |hv: &ScBytes| {
-            let mut vnew: Vec<u8> = hv.metered_clone(&self.0.budget)?.into();
-            // Passing `len()` since worse case can cause reallocation.
-            self.charge_budget(CostType::BytesPush, hv.len() as u64)?;
+            let mut vnew: Vec<u8> = hv.metered_clone(self.as_budget())?.into();
+            // `Vec::push` may inccur reallocation if capacity is reached. Here we charge the cost
+            // of deep-cloning`len` bytes as the worst case. The actual bytes reallocated (if it
+            // happens) will exceed `len` but we don't know the factor (that is internal).
+            <Vec<u8> as MeteredClone>::charge_deep_clone(&vnew, self.as_budget())?;
```

Comment:
At least in part because we're storing the host object as ScBytes, to make conversion-out closer to a no-op. Also because I suspect the iterator-chaining stuff that's used to clone MeteredVector won't optimize quite as well as a byte-slice-copy. But it might be worth checking both of those assumptions. Maybe file a followup bug?

Reply:
I see. But from the requirements standpoint they're the same, right? (i.e. both are immutable vectors and Bytes is conceptually a 'specialization' of `MeterecVec<u8>`). If so, we should make sure both have consistent metering and bytes is not more expensive than a generic vec.

Reply:
If we were to use a `MeteredVec` here, we would first clone the bytes from ScBytes into a `MeteredVec`, then call `push` on it which will clone the whole vector with 1 extra element in it, this will be `O(n^2)`.
Here we just clone it into a Vec<u8> and call push on it, which is just `O(n)`.

Reply:
O(n) copy + O(n) push is still O(n), right? (as long as you just use `MeteredVec` as your container. I'm ok with the current implementation as well, just trying to figure out if we could simplify this and (more importantly) that metering is consistent for the similar data structures.